### PR TITLE
Fix user list infinite loading state in user settings

### LINKED ISSUE #25133: 20.0.5 - groups display at buttom "Keine Benutzer vorhanden" even if there are users
### Steps to reproduce
when i go through my user-groups at "https://xyz/index.php/settings/users" some groups show "Keine Benutzer vorhanden" at the bottom even when there are users in the groups.

### Expected behaviour
"Keine Benutzer vorhanden" should not be there.

### Actual behaviour
"Keine Benutzer vorhanden" get displayed at button even if there are users.

**Nextcloud version:** 20.0.5

### PR DESCRIPTION
**Pull request by julien-nc**
Fix https://github.com/nextcloud/server/issues/25133
This is the same as #25195 but rebased on master with resolved conflicts.

"No users in here" message is shown when there actually are displayed users.

This is because `$state.loaded()` is not called when the retrieved user list length is less than the 25 limit. It has to be called at least once otherwise vue-infinite-loading shows the `no-results` slot.

`$state.loaded()` is now called if we actually fetch users.

There is still one problem with this approach:
If we receive users **but** the displayed filtered list is empty, the `no-results` slot will not be displayed because we called `$state.loaded()`.
We could face this problem with the "Disabled" filter but thankfully the related navigation item does not appear when there is no disabled user.
But it might appear in the future with a new filter that can be applied while it leads to an empty displayed user list.

### CODE REVIEW

File: apps/settings/src/components/UserList.vue [hunk 1]
```diff
@@ -472,7 +472,14 @@ export default {
 				search: this.searchQuery,
 			})
 				.then((response) => {
-					response ? $state.loaded() : $state.complete()
+					// we loaded something
+					if (this.filteredUsers.length > 0) {
+						$state.loaded()
+					}
+					// no more results
+					if (!response) {
```
Comment (PVince81) [inline apps/settings/src/components/UserList.vue:480]: what if filteredUsers has zero entries due to the filter but the response was valid ?
as far as I can see there's some client side filtering happening also in the computed property `filteredUsers`

**Review by PVince81 — DISMISSED** [review]

File: apps/settings/src/components/UserList.vue (hunk 1 above)
Comment (eneiluj) [reply apps/settings/src/components/UserList.vue:480]: I think it works fine in the case you mention. If `filteredUsers` is empty then `$state.loaded()` is not called but the next user list request (with offset) is still done. `$state` is specific to the infinite loading component, it won't stop the request loop.

Checking if `filteredUsers` computed prop is empty is exactly what we need to determine if we should display the empty content message or not.

`(!response)` is true when request result is empty which means we received all users.

**Comment by rullzer** [comment]
master is 22. Backport once merge if needed.

**Comment by szaimen** [comment]
Any update here?

**Comment by eneiluj** [comment]
@PVince81 Did you get a chance to test this PR and https://github.com/nextcloud/server/pull/25195 to check the difference?

**Comment by eneiluj** [comment]
@PVince81 You're right, #25195 works better than my solution (except one small remaining problem mentioned at the end of this PR's description).

So this PR is now the same as #25195, rebased on master, conflicts resolved.

File: apps/settings/src/store/users.js [hunk 2]
```diff
@@ -224,11 +224,12 @@ const actions = {
 				cancelToken: searchRequestCancelSource.token,
 			})
 				.then((response) => {
-					if (Object.keys(response.data.ocs.data.users).length > 0) {
+					const usersCount = Object.keys(response.data.ocs.data.users).length
+					if (usersCount > 0) {
 						context.commit('appendUsers', response.data.ocs.data.users)
-						return Object.keys(response.data.ocs.data.users).length >= limit
+						return usersCount
```
Comment (Pytal) [inline apps/settings/src/store/users.js:230]:
```suggestion
```

File: apps/settings/src/store/users.js [hunk 3]
```diff
@@ -224,11 +224,12 @@ const actions = {
...
 					}
-					return false
+					return 0
```
Comment (Pytal) [inline apps/settings/src/store/users.js:232]:
```suggestion
					return usersCount
```
Combined with the above makes it clearer :)

File: apps/settings/src/store/users.js [hunk 4]
```diff
@@ -241,11 +242,12 @@ const actions = {
 			cancelToken: searchRequestCancelSource.token,
 		})
 			.then((response) => {
-				if (Object.keys(response.data.ocs.data.users).length > 0) {
+				const usersCount = Object.keys(response.data.ocs.data.users).length
+				if (usersCount > 0) {
 					context.commit('appendUsers', response.data.ocs.data.users)
-					return Object.keys(response.data.ocs.data.users).length >= limit
+					return usersCount
```
Comment (Pytal) [inline apps/settings/src/store/users.js:248]:
```suggestion
```

File: apps/settings/src/store/users.js [hunk 5]
```diff
@@ -241,11 +242,12 @@ const actions = {
...
 				}
-				return false
+				return 0
```
Comment (Pytal) [inline apps/settings/src/store/users.js:250]:
```suggestion
				return usersCount
```
Same :)

**Comment by skjnldsv** [comment]
Let's get https://github.com/nextcloud/server/pull/28202 in before

**Comment by eneiluj** [comment]
@Pytal Thanks for the review. Fixed, rebased on master and pushed.

@skjnldsv Yes sure. I'll ask the bot to compile when #28202 is in.

**Review by Pytal — APPROVED** [review]

**Comment by eneiluj** [comment]
I don't know why `/drone/src/tests/acceptance/features/users.feature:48` scenario is failing. I followed it manually on a local test instance and it works fine...

**Comment by szaimen** [comment]
@benediktg please rebase

**Comment by eneiluj** [comment]
@szaimen done

**Review by szaimen — APPROVED** [review]
LGTM :)

**Comment by eneiluj** [comment]
I don't know why but this bug does not affect stable20 although the implementation is the same. Anyway, no need to backport to stable20 :grin:.

**Comment by szaimen** [comment]
Thanks for the backports! :)

**Pull request merged**